presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java

```diff
@@ -13,114 +13,172 @@
     private static final String TASK_HASH_BUILD_CONCURRENCY = "task_hash_build_concurrency";
     private static final String TASK_AGGREGATION_CONCURRENCY = "task_aggregation_concurrency";
-    private static final String QUERY_MAX_MEMORY = "query_max_memory";
+    public static final String QUERY_MAX_MEMORY = "query_max_memory";
     private static final String REDISTRIBUTE_WRITES = "redistribute_writes";
 
-    private SystemSessionProperties() {}
+    private final List<SessionPropertyMetadata<?>> sessionProperties;
 
-    private static boolean isEnabled(String propertyName, Session session, boolean defaultValue)
+    public SystemSessionProperties()
     {
-        String enabled = session.getSystemProperties().get(propertyName);
-        if (enabled == null) {
-            return defaultValue;
-        }
-
-        return Boolean.valueOf(enabled);
+        this(new QueryManagerConfig(), new TaskManagerConfig(), new MemoryManagerConfig(), new FeaturesConfig());
     }
 
-    private static int getNumber(String propertyName, Session session, int defaultValue)
+    @Inject
+    public SystemSessionProperties(
+            QueryManagerConfig queryManagerConfig,
+            TaskManagerConfig taskManagerConfig,
+            MemoryManagerConfig memoryManagerConfig,
+            FeaturesConfig featuresConfig)
     {
-        String count = session.getSystemProperties().get(propertyName);
-        if (count != null) {
-            try {
-                return Integer.parseInt(count);
-            }
-            catch (NumberFormatException ignored) {
-            }
-        }
-
-        return defaultValue;
+        sessionProperties = ImmutableList.of(
+                booleanSessionProperty(
+                        OPTIMIZE_HASH_GENERATION,
+                        "Compute hash code for distribution, joins, and aggregations early in query plan",
+                        featuresConfig.isOptimizeHashGeneration(),
+                        false),
+                booleanSessionProperty(
+                        DISTRIBUTED_JOIN,
+                        "Use a distributed join instead of a broadcast join",
+                        featuresConfig.isDistributedJoinsEnabled(),
+                        false),
+                integerSessionProperty(
+                        HASH_PARTITION_COUNT,
+                        "Number of partitions for distributed joins and aggregations",
+                        queryManagerConfig.getInitialHashPartitions(),
+                        false),
+                booleanSessionProperty(
+                        PREFER_STREAMING_OPERATORS,
+                        "Prefer source table layouts that produce streaming operators",
+                        false,
+                        false),
+                integerSessionProperty(
+                        TASK_WRITER_COUNT,
+                        "Default number of local parallel table writer jobs per worker",
```

Comment:
Nevermind, I see the "jobs" phrasing in the other descriptions. I will try to think of a better word or phrasing.